Add a user_data provisioning type (http://theforeman.org/issues/1561)

pccowboy opened the pull request:
This has been lightly tested, with a provisioning template of 

```
#cloud-config
repo_releasever: 2011.09
```

I will testing it more on my development puppet box over the next week, I am hoping you find it useful.

Not sure why there are 8 commits, though - you already merged those.  Sorry for the trouble.

Commit:
fix error in validation for new compute hosts

add backtrace to compute resource failure msg

add user_data provisioning script type.  Add user_data to the creation of a compute instance

wip on user data feature

configTemplate returns nil if a template is not found

Generate and send user-data to an EC2 instance.  Prevent SSH-events from queuing when there is not a finish script.  Make sure any instance changes between queue and post_queue get saved.

taking a hint from app/models/host.rb, make sure that our save after getting an IP doesnt start validation all over again

adjust the validation method, so that we can use a simple check to prevent queuing tasks when a finish script is not present

only force presence of a finish script on non-EC2 provider

Commit:
fix error in validation for new compute hosts

Commit:
fix error in validation for new compute hosts

add backtrace to compute resource failure msg

add user_data provisioning script type.  Add user_data to the creation of a compute instance

wip on user data feature

configTemplate returns nil if a template is not found

Generate and send user-data to an EC2 instance.  Prevent SSH-events from queuing when there is not a finish script.  Make sure any instance changes between queue and post_queue get saved.

taking a hint from app/models/host.rb, make sure that our save after getting an IP doesnt start validation all over again

adjust the validation method, so that we can use a simple check to prevent queuing tasks when a finish script is not present

only force presence of a finish script on non-EC2 provider

Commit:
fix error in validation for new compute hosts

Commit:
fix error in validation for new compute hosts

add backtrace to compute resource failure msg

add user_data provisioning script type.  Add user_data to the creation of a compute instance

wip on user data feature

configTemplate returns nil if a template is not found

Generate and send user-data to an EC2 instance.  Prevent SSH-events from queuing when there is not a finish script.  Make sure any instance changes between queue and post_queue get saved.

taking a hint from app/models/host.rb, make sure that our save after getting an IP doesnt start validation all over again

adjust the validation method, so that we can use a simple check to prevent queuing tasks when a finish script is not present

only force presence of a finish script on non-EC2 provider

Commit:
fix error in validation for new compute hosts

add backtrace to compute resource failure msg

add user_data provisioning script type.  Add user_data to the creation of a compute instance

wip on user data feature

configTemplate returns nil if a template is not found

Generate and send user-data to an EC2 instance.  Prevent SSH-events from queuing when there is not a finish script.  Make sure any instance changes between queue and post_queue get saved.

taking a hint from app/models/host.rb, make sure that our save after getting an IP doesnt start validation all over again

adjust the validation method, so that we can use a simple check to prevent queuing tasks when a finish script is not present

only force presence of a finish script on non-EC2 provider

Commit:
fix error in validation for new compute hosts

Commit:
debugging, to get the whole message

Commit:
debugging, to get the whole message

Commit:
debugging, to get the whole message

Commit:
debugging, to get the whole message

Commit:
merge updated fix for compute resource not needing partition table

Commit:
add backtrace to compute resource failure msg

Commit:
add backtrace to compute resource failure msg

Commit:
add backtrace to compute resource failure msg

Commit:
add debug logging for Fog errors on create

Commit:
add debug logging for Fog errors on create

Commit:
add debug logging for Fog errors on create

Commit:
retry on host not available, add debugging.

Commit:
correct merge conflict (why?)

Commit:
add user_data provisioning script type.  Add user_data to the creation of a compute instance

Commit:
add user_data provisioning script type.  Add user_data to the creation of a compute instance

Commit:
add user_data provisioning script type.  Add user_data to the creation of a compute instance

Commit:
add user_data provisioning script type.  Add user_data to the creation of a compute instance

Commit:
add user_data provisioning script type.  Add user_data to the creation of a compute instance

iNecas committed:
fixes #1463 - let Foreman run on a sub path

e.g. for it being available on http://localhost:3000/foreman you run:

    RAILS_RELATIVE_URL_ROOT=/foreman rails s

ohadlevy commented:
how do we handle certs this way?

pccowboy commented:
I am currently using auto-sign, but cloud-config has a way to send the 
cert, detailed at 
http://bazaar.launchpad.net/~cloud-init-dev/cloud-init/trunk/view/head:/doc/examples/cloud-config-ca-certs.txt

#cloud-config
#

# This is an example file to configure an instance's trusted CA certificates

# system-wide for SSL/TLS trust establishment when the instance boots

for the

# first time.

#

# Make sure that this file is valid yaml before starting instances.

# It should be passed as user-data when starting the instance.

ca-certs:
   # If present and set to True, the 'remove-defaults' parameter will remove
   # all the default trusted CA certificates that are normally shipped with
   # Ubuntu.
   # This is mainly for paranoid admins - most users will not need this
   # functionality.
   remove-defaults: true

   # If present, the 'trusted' parameter should contain a certificate 
(or list
   # of certificates) to add to the system as trusted CA certificates.
   # Pay close attention to the YAML multiline list syntax.  The example 
shown
   # here is for a list of multiline certificates.
   trusted:
- |
  -----BEGIN CERTIFICATE-----
  YOUR-ORGS-TRUSTED-CA-CERT-HERE
  -----END CERTIFICATE-----
- |
  -----BEGIN CERTIFICATE-----
  YOUR-ORGS-TRUSTED-CA-CERT-HERE
  -----END CERTIFICATE-----

Ohad Levy wrote:

> how do we handle certs this way?
> 
> ---
> 
> Reply to this email directly or view it on GitHub:
> https://github.com/theforeman/foreman/pull/67#issuecomment-6244802
> 
> .

Commit:
Merge branch 'develop' of https://github.com/theforeman/foreman into develop

Commit:
wip on user data feature

Commit:
wip on user data feature

Commit:
wip on user data feature

Commit:
wip on user data feature

Commit:
Merge branch 'develop' of https://github.com/theforeman/foreman into develop

Commit:
configTemplate returns nil if a template is not found

Commit:
configTemplate returns nil if a template is not found

Commit:
configTemplate returns nil if a template is not found

Commit:
configTemplate returns nil if a template is not found

Commit:
configTemplate returns nil if a template is not found

Commit:
merge conflict

pccowboy commented:
I found what I think is a bug in a commit from abenari - commit https://github.com/pccowboy/foreman/commit/5051c7a00bc31325825e62794157c6a3183c32f6 has a (poor) fix.

ohadlevy commented on app/models/orchestration/ssh_provision.rb:
```diff
@@ -92,7 +92,8 @@ def validate_ssh_provisioning
     begin
-      configTemplate(:kind => "finish")
+      template = configTemplate(:kind => "finish")
+      status = (template != nil)
```
yep, however I cant merge your current pull request, either you create another one, or I'll cherry pick the commit from here.

Commit:
Generate and send user-data to an EC2 instance.  Prevent SSH-events from queuing when there is not a finish script.  Make sure any instance changes between queue and post_queue get saved.

Commit:
Generate and send user-data to an EC2 instance.  Prevent SSH-events from queuing when there is not a finish script.  Make sure any instance changes between queue and post_queue get saved.

Commit:
Merge branch 'develop' of https://github.com/theforeman/foreman into develop

Commit:
Merge branch 'ec2' of https://github.com/pccowboy/foreman into ec2

ohadlevy commented:
i meant puppet ca, but, do you think its considered a trusted source to put the  certificate private key in exiternal userdata / metadata service?

pccowboy commented:
In the case of EC2, I think so, since all communication to AWS is encrypted, and the userdata can only get to your instance if you have the correct AWS security credentials.  I will ask their support staff as well, just to make sure that the userdata is not logged anywhere else at all.

Commit:
Merge branch 'develop' of https://github.com/theforeman/foreman into develop

Commit:
Merge branch 'develop' of https://github.com/theforeman/foreman into develop

Commit:
Update gemfile to rails 3.0.15, for mysql users, per https://github.com/rails/rails/issues/6717

Commit:
Update gemfile to rails 3.0.15, for mysql users, per https://github.com/rails/rails/issues/6717

Commit:
update provisioning failure msg when no finish or user_data templates were found.  This should be expanded to a better check based upon the provider type, IMO.  But, that is work for another day.

Commit:
update provisioning failure msg when no finish or user_data templates were found.  This should be expanded to a better check based upon the provider type, IMO.  But, that is work for another day.

Commit:
update provisioning failure msg when no finish or user_data templates were found.  This should be expanded to a better check based upon the provider type, IMO.  But, that is work for another day.

Commit:
Merge branch 'develop' of https://github.com/theforeman/foreman into develop

Commit:
merge two files

logicminds committed:
Merge pull request #89 from logicminds/1710

fixes #1710 - Hosts json index function returns too much

ohadlevy committed:
refs #1586 - no need to copy hostgroup root pw

the idea here is not to copy the pw from the hostgroup, as if
you would like to change the hostgroup password, it would change it
to all systems which are using the hostgroup pw (and did not override it manually).

ohadlevy committed:
fixes #1686 - can't choose subnet when domain comes from hostgroup

ohadlevy committed:
update to latest released version of fog

ohadlevy committed:
fixes #1707 - db:migrate fail on in bookmark.rb with ruby1.9

ohadlevy committed:
ensure that migration script is a bit more rubost for audit uses, refs #1700

(sadly the audited gem alias the .user method, and sometime returns
strings and other times return a user object)

ohadlevy committed:
facter sometime send wrong mac address with bonded nics,

This patch is kind of a workaround to ensure we dont override valueable information.

ohadlevy committed:
refs #1586 - ensures that hostgroup hierarchy order is processed correctly

ohadlevy committed:
added commit comment, history and preview to the template editor.

ohadlevy committed:
1.0RC2

ohadlevy committed:
minor fix and added a few tests to ensure config templates works correctly

ohadlevy committed:
fixes #1714 allowing users with build permissions to access the submit multiple

ohadlevy committed:
fixes #1715 - rake puppet:import:puppet_classes[batch] stopped working

ohadlevy committed:
fixes #1638 - remove whitespaces for user email instead of validation error

abenari committed:
audit time ago in words.

abenari committed:
fixes #1712 When Ec2 provisioning, resolve button shouldn't show pxe nor provisioning template.

abenari committed:
fixes #1677 - Fact pie chart alters fact strings to lower case

ohadlevy committed:
dont fail if console bundler group is not installed

abenari committed:
fixes #1563 problem cloning host in latest develop

Commit:
Fixed migrations for mysql2

frimik committed:
fixes #1716 - remove whitespace from hostname

ohadlevy committed:
disk layout should really not be unique.

ohadlevy committed:
fixed wrong test - refs #1677

ohadlevy committed:
missing syntax

abenari committed:
top navigation bar looks correct on tablet and smart-phone screen size.

ohadlevy committed:
make sure that info messages are not shown when filtering is enabled refs #1634

abenari committed:
fixes #1240 more menu fixed for users with no permitions.

abenari committed:
host list page fixed for smart-phones and tablet screen sizes.

ekohl committed:
Return a HTTP 500 when we fail on handle_ca (fixes #1725)

Commit:
Add Archlinux logo back to hosts page

Commit:
Figure out how to make the power buttons work properly.  Update the IP address when the power state changes

ohadlevy committed:
avoid breaking mysql migration from an empty db

abenari committed:
fixes the more menu bug

abenari committed:
fixes #1726 ie support for edit template (with a simple text-box).

abenari committed:
fixes #1165 add inherited params display and override option to the host form.

Signed-off-by: Ohad Levy <ohadlevy@gmail.com>

abenari committed:
fixes more menu issues

ohadlevy committed:
fixes parameters whitespace validations

ohadlevy committed:
1.0RC3 released

abenari committed:
fixes #1729 safer migration code.

ohadlevy committed:
refs #1589, ensure that vmware public key is automaticilly imported when testing connection

abenari committed:
fixes #1730 broken ui when errors and help inline.

ohadlevy committed:
ensures that compute resources are shown in the host form only if provisioning is enabled

ohadlevy committed:
fixes #1743 - auditing RC2 breaks foreman, forcing RC1 for now

ohadlevy committed:
fixes #1737 - make sure mac based matching are case insensetive

abenari committed:
use file reader for the config template file upload.

abenari committed:
fixed dropdown location for action button and bookmark.

ohadlevy committed:
remove unneded hidden links in the more tab

ohadlevy committed:
[UI] - added missing label classes

ohadlevy committed:
adds missing EC2 propetries view

GregSutcliffe committed:
Update first-run splash page with a note about Smart Proxies, fixes #1721

abenari committed:
fixes #1746 selecting a cluster in oVirt remove subnet and ip from the network tab in new/edit host

abenari committed:
fixes #1747 host selected puppet classes are afected by the classes filter.

ohadlevy committed:
[UI] - convert all bulk operations into buttons

ohadlevy committed:
dont enforce host domain validations on non managed hosts

ohadlevy committed:
[UI] convert all titles to tooltips

Commit:
explicitly add name to create_vm call, so that the entered hostname takes precedence over the foreman-generated name

Commit:
how did I get two merge conflicts again?

Commit:
fix a badly botched merge

Commit:
add user_data example scripts and a migration to add them.

pccowboy commented:
I am going to close this request, and open it after a vigorous rebase.
